server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -91,6 +93,7 @@ public class SinkQuerySegmentWalker implements QuerySegmentWalker
   private final String dataSource;
 
   private final VersionedIntervalTimeline<String, Sink> sinkTimeline;
+  private final VersionedIntervalTimeline<String, SegmentDescriptorOvershadowable> upgradeDescriptorTimeline;
```

Comment:
Let's call it `upgradedSegmentsTimeline` instead, descriptor is a little confusing.
Also add a javadoc mentioning that the purpose of this timeline is to lookup the ID/interval of an upgraded segment which is then used to find the base segment using the map.